File: lab/js/sim/mapgen.js
// sim/mapgen.js — a Customize-New-World map generator following func_064A10's
// passes P0–P6 (spec/systems/map_generation.md). The PASS STRUCTURE, terrain
// immediates, climate→terrain tables, landmass target, and border rules are
// BYTE_VERIFIED (cited inline). The RNG and the exact blob-walk / smoothing fill
// conditions are NOT byte-pinned, so the OUTPUT is MODELED (R): a different map per
// seed with the right structure, not a byte-exact reproduction of a DOS seed.
//
// Output tiles use the SAME encoding the committed AMER2 grid + the M3 compositor
// read: base id in bits0-4 (0..7 land, 8..15 forested, 24 Arctic, 25 Ocean,
// 26 Sea Lane), hills bit 0x20, mountain bit 0x20|0x80, river bit 0x40 — so a
// generated map composites identically through sim/mapview.js.

const OCEAN = 25, SEA_LANE = 26, ARCTIC = 24;

// Climate band → base terrain id (BYTE_VERIFIED inline jump tables @0x64CFC / @0x6504E).
const CLIMATE_N = [5, 4, 1, 3, 2, 2];   // Savannah,Grassland,Desert,Prairie,Plains,Plains
const CLIMATE_S = [2, 3, 3, 4, 6, 7];   // Plains,Prairie,Prairie,Grassland,Marsh,Swamp

// Seedable PRNG (mulberry32) — MODELED stand-in for the DOS LCG (0x181F:0x4D4).
function rng(seed) {
  let a = seed >>> 0;
  return () => {
    a |= 0; a = (a + 0x6D2B79F5) | 0;
    let t = Math.imul(a ^ (a >>> 15), 1 | a);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

// Smooth value-noise field (seeded), bilinearly interpolated over a coarse grid —
// used to make climate boundaries wavy (not even strips) and to cluster elevation
// into ranges. NOT byte-derived: the DOS generator's perturbation isn't pinned, so
// this is part of the modeled (R) randomization. cell = grid spacing in tiles.
function noiseField(seed, cell) {
  const h = (xi, yi) => {                       // hash lattice point → [0,1)
    let a = (xi * 374761393 + yi * 668265263 + seed * 2246822519) >>> 0;
    a = Math.imul(a ^ (a >>> 13), 1274126177) >>> 0;
    return ((a ^ (a >>> 16)) >>> 0) / 4294967296;
  };
  const sm = (t) => t * t * (3 - 2 * t);        // smoothstep
  return (x, y) => {
    const gx = x / cell, gy = y / cell;
    const x0 = Math.floor(gx), y0 = Math.floor(gy);
    const fx = sm(gx - x0), fy = sm(gy - y0);
    const a = h(x0, y0), b = h(x0 + 1, y0), c = h(x0, y0 + 1), d = h(x0 + 1, y0 + 1);
    return (a * (1 - fx) + b * fx) * (1 - fy) + (c * (1 - fx) + d * fx) * fy;  // [0,1)
  };
}

// The four byte-verified Customize-New-World parameters (DGROUP:0x1E7E, mod-3 0..2),
// with their exact in-game menu labels (GAME.TXT @CLAND/@CCONT/@CTEMP/@CCLIM):
export const CUSTOMIZE = {
  landMass:    { key: '@CLAND', title: 'LAND MASS',   options: ['Small', 'Normal', 'Large'] },
  landForm:    { key: '@CCONT', title: 'LAND FORM',   options: ['Archipelago', 'Normal', 'Large Continents'] },
  temperature: { key: '@CTEMP', title: 'TEMPERATURE', options: ['Cool', 'Temperate', 'Warm'] },
  climate:     { key: '@CCLIM', title: 'CLIMATE',     options: ['Arid', 'Normal', 'Wet'] },
};

// params: { landMass, landForm, temperature, climate, seed }, each Customize var 0..2
// (idx0=land-mass p1, idx1=land-form p2, idx2=temperature, idx3=moisture/climate —
//  DGROUP:0x1E7E, written mod-3 by the Customize dialog func_070060 @0x701AD).
export function generate(params = {}) {
  const W = 58, H = 72;                 // random-map defaults @0x75702/@0x75708 (0x3A×0x48) [B]
  const landMass = clamp(params.landMass ?? 1, 0, 2);       // idx0 p1
  const landForm = clamp(params.landForm ?? 1, 0, 2);       // idx1 p2
  const temperature = clamp(params.temperature ?? 1, 0, 2); // idx2 (Cool..Warm)
  const climateV = clamp(params.climate ?? 1, 0, 2);        // idx3 moisture (Arid..Wet)
  // Map the named vars onto the generator's internals:
  const p1 = landMass + 1, p2 = landForm;                   // landmass target (p1+p2+1)·0x140 @0x64AAD
  const climateShift = 1 - temperature;                     // Cool → poleward bands, Warm → equatorial
  const moisture = climateV;                                // Arid(0)/Normal(1)/Wet(2)
  const seed = (params.seed ?? 1) >>> 0;
  const rand = rng(seed);
  const ri = (lo, hi) => { lo = Math.floor(lo); hi = Math.floor(hi); return lo + Math.floor(rand() * (hi - lo + 1)); };

  const t = new Uint8Array(W * H).fill(OCEAN);   // P0: all Ocean (0x19) @0x64A4B
  const land = new Uint8Array(W * H);            // landmass mask (P1)
  const at = (x, y) => (x < 0 || y < 0 || x >= W || y >= H ? -1 : y * W + x);

  // --- P1 landmass: blob-growth random walk; target ≈(p1+p2+1)·0x140 (@0x64AAD) [B target] ---
  const target = (p1 + p2 + 1) * 0x140;
  const DX = [0, 1, 1, 1, 0, -1, -1, -1], DY = [-1, -1, 0, 1, 1, 1, 0, -1];   // 8-dir compass (DS:0xB4/0xBE)
  let placed = 0;
  // LAND FORM controls coherence: Archipelago = many scattered blobs, Large
  // Continents = few big ones (fewer walkers over the same land target).
  const nWalk = landForm === 0 ? 6 + landMass : landForm === 1 ? 3 : 2;
  // Round-robin drunkard's walk over a few walkers until we hit the land target.
  const walkers = [];
  for (let s = 0; s < nWalk; s++) walkers.push({ x: ri(W * 0.25, W * 0.7), y: Math.floor((H / (nWalk + 1)) * (s + 1)) });
  let iter = 0; const maxIter = target * 60;
  while (placed < target && iter++ < maxIter) {
    for (const w of walkers) {
      if (placed >= target) break;
      const i = at(w.x, w.y);
      if (i >= 0 && !land[i]) { land[i] = 1; placed++; }
      // 4-neighbour land mask (mask 6/9 ⇒ fill the boxed-in cell)
      let m = 0;
      if (at(w.x, w.y - 1) >= 0 && land[at(w.x, w.y - 1)]) m |= 8;
      if (at(w.x, w.y + 1) >= 0 && land[at(w.x, w.y + 1)]) m |= 4;
      if (at(w.x - 1, w.y) >= 0 && land[at(w.x - 1, w.y)]) m |= 2;
      if (at(w.x + 1, w.y) >= 0 && land[at(w.x + 1, w.y)]) m |= 1;
      if ((m === 6 || m === 9) && i >= 0 && !land[i]) { land[i] = 1; placed++; }
      const d = ri(0, 7);
      w.x = clamp(w.x + DX[d], 1, W - 3);     // keep off the sea-lane columns
      w.y = clamp(w.y + DY[d], 1, H - 2);
    }
  }

  // --- P2 climate: latitude band → base terrain; set hills/forest (@0x64CF6/@0x65048) [tables B] ---
  // The latitude→band mapping is byte-verified; the DOS generator does NOT lay even
  // strips, so we perturb the band with a smooth noise field (modeled R) — boundaries
  // wander, and an independent elevation field clusters hills/mountains into ranges.
  const climateNoise = noiseField(seed ^ 0x1234, 9);     // ~9-tile climate cells → wavy bands
  const localNoise = noiseField(seed ^ 0x9E37, 3);       // fine per-tile terrain jitter
  const elevNoise = noiseField(seed ^ 0x5151, 7);        // elevation → mountain ranges/foothills
  for (let y = 0; y < H; y++) {
    const north = y < H / 2;
    const distFromEq = north ? (H / 2 - y) / (H / 2) : (y - H / 2) / (H / 2);
    for (let x = 0; x < W; x++) {
      const i = y * W + x;
      if (!land[i]) continue;
      // band wanders ±~2 with the climate noise instead of a clean latitude step
      const jitter = (climateNoise(x, y) - 0.5) * 4.2;
      let band = clamp(Math.round(distFromEq * 5 + jitter) + climateShift, 0, 5);
      let base = (north ? CLIMATE_N : CLIMATE_S)[band];
      // local jitter: sometimes borrow an adjacent band's terrain so a band isn't one flat colour
      const lj = localNoise(x, y);
      if (lj > 0.86 && band < 5) base = (north ? CLIMATE_N : CLIMATE_S)[band + 1];
      else if (lj < 0.14 && band > 0) base = (north ? CLIMATE_N : CLIMATE_S)[band - 1];
      if (!north && (base === 6 || base === 7) && moisture < 1) base -= 2;   // dry: Marsh/Swamp → drier (@moisture −2)
      let b = base & 0x1F;
      // elevation from the clustered field: high zones = mountain spines, flanks = hills
      const e = elevNoise(x, y);
      if (e > 0.78) b |= 0x20 | 0x80;        // mountain (ranges, not specks)
      else if (e > 0.66) b |= 0x20;          // hills (foothills around the spines)
      t[i] = b;
    }
  }

  // --- P3 smoothing: fold unforested→forested (+8, @0x653F8) in NOISE PATCHES (not
  // speckle); de-speckle interior ocean. forestThresh ↓ with moisture ⇒ wetter = more forest. ---
  const forestNoise = noiseField(seed ^ 0x7A21, 5);   // ~5-tile forest patches
  const forestThresh = 0.62 - moisture * 0.12;        // Arid 0.62 / Normal 0.50 / Wet 0.38
  for (let y = 0; y < H; y++) for (let x = 0; x < W; x++) {
    const i = y * W + x;
    if (!land[i]) continue;
    const baseId = t[i] & 0x1F;
    if (baseId < 8 && (t[i] & 0x20) === 0 && baseId !== 1 && forestNoise(x, y) > forestThresh)
      t[i] = (t[i] & 0xE0) | ((baseId & 7) | 8);   // forested variant (band 8..15)
  }
  // remove 1-tile ocean holes surrounded by land
  for (let y = 1; y < H - 1; y++) for (let x = 1; x < W - 1; x++) {
    const i = y * W + x;
    if (land[i]) continue;
    const ln = land[at(x, y - 1)] + land[at(x, y + 1)] + land[at(x - 1, y)] + land[at(x + 1, y)];
    if (ln === 4) { land[i] = 1; t[i] = 2; }       // fill with Plains
  }

  // --- P4 rivers: a few interior→coast paths carry the river bit 0x40 (@0x65BC2) ---
  const rivers = Math.max(2, p1 + p2 + 2);
  for (let r = 0; r < rivers; r++) {
    // find a land interior start (retry a handful of times)
    let x = 0, y = 0, found = false;
    for (let tryi = 0; tryi < 40 && !found; tryi++) {
      x = ri(4, W - 6); y = ri(4, H - 6);
      if (land[at(x, y)]) found = true;
    }
    if (!found) continue;
    let guard = 0;
    while (guard++ < 200) {
      const i = at(x, y);
      if (i < 0 || !land[i]) break;        // reached coast/edge
      t[i] |= 0x40;
      x += x > W / 2 ? 1 : -1;             // drift toward the nearest horizontal edge
      y += ri(-1, 1);
      y = clamp(y, 1, H - 2);
    }
  }

  // --- P5 borders: top/bottom rows → Arctic (24), THEN right two columns → Sea Lane
  // (26) so the east edge stays an unbroken pole-to-pole lane (@0x65941/@0x6582A) [B] ---
  for (let x = 0; x < W; x++) { t[x] = ARCTIC; t[(H - 1) * W + x] = ARCTIC; land[x] = 0; land[(H - 1) * W + x] = 0; }
  for (let y = 0; y < H; y++) { t[y * W + (W - 1)] = SEA_LANE; t[y * W + (W - 2)] = SEA_LANE; land[y * W + (W - 1)] = 0; land[y * W + (W - 2)] = 0; }

  // --- P6 starts: 4 powers, y = (H/5)·(p+1) band, x walked inland from the east sea lane (@0x65C9C) [B rule] ---
  const starts = [];
  for (let p = 0; p < 4; p++) {
    const sy = Math.floor((H / 5) * (p + 1));
    let sx = W - 3;
    while (sx > 2 && !land[sy * W + sx]) sx--;     // walk inland until land
    starts.push({ power: p, x: sx, y: sy });
  }

  return { width: W, height: H, tiles: t, starts, landCount: placed, target };
}

function clamp(v, lo, hi) { return v < lo ? lo : v > hi ? hi : v; }

File: viceroy_cpp/sim/mapgen.cpp
// sim/mapgen.cpp -- see mapgen.hpp for the byte citations (func_064A10).
#include "mapgen.hpp"

namespace vc::sim {

namespace {
// The generator's 8-dir compass tables (DS:0xB4 dx / 0xBE dy).
const int DX8[8] = {1, 1, 0, -1, -1, -1, 0, 1};
const int DY8[8] = {0, 1, 1, 1, 0, -1, -1, -1};
}

void generate_map(World& w, const MapGenParams& p, const RandFn& rng,
                  MapStart starts[4]) {
    const int W = MAP_W, H = MAP_H;                 // 58x72 (@0x75702/@0x75708)
    w.map_w = W; w.map_h = H;
    w.terrain.assign((size_t)W * H, 25);            // P0: all-Ocean fill (0x19 @0x64A4B)
    w.improve.clear(); w.fog.clear();
    auto at = [&](int x, int y) -> uint8_t& { return w.terrain[(size_t)y * W + x]; };
    auto in = [&](int x, int y) { return x >= 2 && x < W - 3 && y >= 2 && y < H - 2; };

    // P1: blob-growth landmass -- (p1+p2+1)*0x140 land tiles grown by 8-dir
    // random walks (@0x64AAD seed count; walk @0x64B5A). A walker restarts on
    // leaving bounds; the fill-mask trigger detail is folded into the walk.
    long budget = (long)(p.land + p.landform + 1) * 0x140;
    while (budget > 0) {
        int x = rng(W / 4, 3 * W / 4), y = rng(3, H - 4);
        for (int step = 0; step < 64 && budget > 0; ++step) {
            if (!in(x, y)) break;
            if ((at(x, y) & 0x1F) == 25) { at(x, y) = 2; --budget; }   // land (Plains stub)
            const int d = rng(0, 7);
            x += DX8[d]; y += DY8[d];
        }
    }

    // P2: latitude climate bands -> base terrain (the byte-verified inline
    // jump tables), then the feature bits. Band = (row-in-hemisphere >> 2),
    // clamped 0..5, warped one band by the temperature parameter. Feature
    // DENSITIES here are RECONSTRUCTED (climate parameter biases forest).
    for (int y = 0; y < H; ++y)
        for (int x = 0; x < W; ++x) {
            if ((at(x, y) & 0x1F) == 25) continue;
            const bool south = y >= H / 2;
            int row = south ? (H - 1 - y) : y;
            int band = (row >> 2) + (p.temperature - 1);
            if (band < 0) band = 0;
            if (band > 5) band = 5;
            int base = climate_base_terrain(band, south);
            if (base == 6 && rng(1, 2) == 2) base = 4;   // the Marsh 50% roll (@0x6500C)
            // features: hills / forest (the 0x20/0x80 board bits ~ our ids/bit6)
            const int r = rng(1, 12);
            if (r == 1)      base = 28;                  // Hills
            else if (r == 2) base = 27;                  // Mountains
            else if (base <= 7 && rng(1, 3) <= 1 + p.climate)   // wetter -> denser forest
                base = (base & 7) | 8;                   // P3 fold: unforested -> +8
            at(x, y) = (uint8_t)base;
        }

    // P3: light smoothing -- stray interior ocean surrounded by >= 6 land
    // becomes open land (the relaxation pass, budget-limited in the EXE).
    for (int y = 1; y < H - 1; ++y)
        for (int x = 1; x < W - 3; ++x) {
            if ((at(x, y) & 0x1F) != 25) continue;
            int n = 0;
            for (int d = 0; d < 8; ++d)
                if ((at(x + DX8[d], y + DY8[d]) & 0x1F) != 25) ++n;
            if (n >= 6) at(x, y) = 2;
        }

    // P4: rivers -- a few walks from high ground toward the sea setting the
    // .MP river bit (bit 5; the runtime board uses 0x40, the FILE packing
    // 0x20 -- our World carries the file packing). Count RECONSTRUCTED.
    for (int r = 0; r < 3 + p.climate; ++r) {
        int x = rng(4, W - 6), y = rng(4, H - 5);
        for (int step = 0; step < 30; ++step) {
            const int t = at(x, y) & 0x1F;
            if (t == 25 || t == 26) break;               // reached the sea
            at(x, y) |= 0x20;                            // river overlay (MP_FORMAT bit 5)
            const int d = rng(0, 7);
            x += DX8[d]; y += DY8[d];
            if (x < 1 || x >= W - 1 || y < 1 || y >= H - 1) break;
        }
    }

    // P5: borders -- top/bottom rows Arctic first (0x18 @0x6582A), THEN the
    // right two columns Sea Lane (0x1A @0x65941/@0x65975; the higher site runs
    // later, so the corners end as Sea Lane).
    for (int x = 0; x < W; ++x) { at(x, 0) = 24; at(x, H - 1) = 24; }
    for (int y = 0; y < H; ++y) { at(W - 1, y) = 26; at(W - 2, y) = 26; }

    // P6: European starts -- y = (H/5)*(p+1) bands (@0x65C9C); x walked inland
    // from the east sea lane to the first land, the ship anchoring on the
    // water tile just east of it.
    for (int pw = 0; pw < 4; ++pw) {
        int y = (H / 5) * (pw + 1);
        if (y < 1) y = 1;
        if (y > H - 2) y = H - 2;
        int sx = W - 3;
        while (sx > 1 && (at(sx, y) & 0x1F) == 25) --sx;
        starts[pw].x = (sx < W - 3) ? sx + 1 : W - 3;    // the water tile east of land
        starts[pw].y = y;
    }
}

} // namespace vc::sim
